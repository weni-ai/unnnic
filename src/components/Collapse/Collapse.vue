<!-- eslint-disable vue/multi-word-component-names -->
<template>
  <div
    :class="[
      'unnnic-collapse',
      `unnnic-collapse--size-${size}`,
      {
        'unnnic-collapse--border-bottom': borderBottom,
        'unnnic-collapse--unspaced-icon': unspacedIcon,
        'unnnic-collapse--active': localActive,
      },
    ]"
  >
    <div
      class="unnnic-collapse__header"
      @click="localActive = !localActive"
    >
      {{ title }}<slot name="header"></slot>

      <UnnnicIcon
        size="xs"
        :icon="localActive ? 'arrow-button-up-1' : 'arrow-button-down-1'"
        :scheme="localActive ? 'neutral-dark' : 'neutral-cloudy'"
      />
    </div>

    <div
      v-show="localActive"
      class="unnnic-collapse__body"
    >
      <slot name="default"></slot>
    </div>
  </div>
</template>

<script setup>
import { ref, watch } from 'vue';
import UnnnicIcon from '../Icon.vue';

const props = defineProps({
  title: String,
  active: Boolean,
  borderBottom: Boolean,
  unspacedIcon: Boolean,
  size: {
    type: String,
    default: 'lg',
    validator(value) {
      return ['md', 'lg'].includes(value);
    },
  },
  modelValue: Boolean,
});

const emit = defineEmits(['update:modelValue', 'change']);

const localActive = ref(false);

watch(localActive, (newActive) => {
  emit('change', newActive);
  emit('update:modelValue', newActive);
});

watch(
  () => props.active,
  (newActive) => {
    localActive.value = newActive;
  },
  { immediate: true },
);

watch(
  () => props.modelValue,
  (newActive) => {
    localActive.value = newActive;
  },
  { immediate: true },
);
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';

.unnnic-collapse {
  padding-bottom: $unnnic-spacing-ant;

  &__header {
    display: flex;
    align-items: center;
    column-gap: $unnnic-spacing-ant;
    justify-content: space-between;
    cursor: pointer;
    padding: $unnnic-spacing-ant 0;
    margin-bottom: -$unnnic-spacing-ant;
    user-select: none;

    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-lg;
    line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
    color: $unnnic-color-neutral-cloudy;

    &:hover {
      color: $unnnic-color-neutral-dark;

      :deep(svg .primary) {
        fill: $unnnic-color-neutral-dark;
      }
    }
  }

  &__body {
    margin-top: $unnnic-spacing-ant;

    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    color: $unnnic-color-neutral-cloudy;
  }

  &--border-bottom {
    padding-bottom: $unnnic-spacing-ant - $unnnic-border-width-thinner;
    border-bottom: $unnnic-border-width-thinner solid
      $unnnic-color-neutral-cleanest;
  }

  &--active .unnnic-collapse__header {
    color: $unnnic-color-neutral-dark;
  }

  &--unspaced-icon .unnnic-collapse__header {
    justify-content: flex-start;
  }

  &--size-md .unnnic-collapse__header {
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
