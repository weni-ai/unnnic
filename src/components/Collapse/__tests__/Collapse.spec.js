import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';
import Collapse from '../Collapse.vue';

describe('Collapse.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Collapse, {
      props: { title: 'Test Title' },
    });
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('Props', () => {
    it('should accept title prop and render it correctly', () => {
      expect(wrapper.find('[data-testid="collapse-header"]').text()).toContain(
        'Test Title',
      );
    });

    it('should accept active prop and set localActive accordingly', async () => {
      await wrapper.setProps({ active: true });
      expect(wrapper.vm.localActive).toBe(true);

      await wrapper.setProps({ active: false });
      expect(wrapper.vm.localActive).toBe(false);
    });

    it('should apply classes based on props', async () => {
      await wrapper.setProps({ borderBottom: true });
      expect(wrapper.classes()).toContain('unnnic-collapse--border-bottom');

      await wrapper.setProps({ unspacedIcon: true });
      expect(wrapper.classes()).toContain('unnnic-collapse--unspaced-icon');

      await wrapper.setProps({ size: 'lg' });
      expect(wrapper.classes()).toContain('unnnic-collapse--size-lg');

      await wrapper.setProps({ size: 'md' });
      expect(wrapper.classes()).toContain('unnnic-collapse--size-md');

      await wrapper.setProps({ size: undefined });
      expect(wrapper.classes()).toContain('unnnic-collapse--size-lg'); // default size

      await wrapper.setProps({ modelValue: true });
      expect(wrapper.classes()).toContain('unnnic-collapse--active');

      await wrapper.setProps({ modelValue: false });
      expect(wrapper.classes()).not.toContain('unnnic-collapse--active');
    });
  });

  describe('Interactions', () => {
    it('should toggle localActive when header is clicked', async () => {
      const header = wrapper.find('[data-testid="collapse-header"]');

      expect(wrapper.vm.localActive).toBe(false);
      await header.trigger('click');
      expect(wrapper.vm.localActive).toBe(true);
      await header.trigger('click');
      expect(wrapper.vm.localActive).toBe(false);
    });

    it('should emit change and update:modelValue on toggle', async () => {
      const header = wrapper.find('[data-testid="collapse-header"]');

      await header.trigger('click');

      expect(wrapper.emitted('change')).toEqual([[true]]);
      expect(wrapper.emitted('update:modelValue')).toEqual([[true]]);

      await header.trigger('click');

      expect(wrapper.emitted('change')).toEqual([[true], [false]]);
      expect(wrapper.emitted('update:modelValue')).toEqual([[true], [false]]);
    });
  });

  describe('Slots', () => {
    it('should render default slot content when active', async () => {
      const defaultSlotContent = 'This is default slot content';
      wrapper = mount(Collapse, {
        props: { title: 'Test Title', active: true },
        slots: { default: defaultSlotContent },
      });
      expect(wrapper.html()).toContain(defaultSlotContent);
    });

    it('should render header slot content if provided', async () => {
      const headerSlotContent =
        '<header data-testid="header-slot">Header Slot Content</header>';
      wrapper = mount(Collapse, {
        props: { title: 'Test Title' },
        slots: { header: headerSlotContent },
      });
      expect(wrapper.find('[data-testid="header-slot"]').exists()).toBe(true);
    });
  });

  describe('Watchers', () => {
    it('should update localActive when active prop changes', async () => {
      expect(wrapper.vm.localActive).toBe(false);
      await wrapper.setProps({ active: true });
      expect(wrapper.vm.localActive).toBe(true);
    });

    it('should update localActive when modelValue prop changes', async () => {
      expect(wrapper.vm.localActive).toBe(false);
      await wrapper.setProps({ modelValue: true });
      expect(wrapper.vm.localActive).toBe(true);
    });
  });
});
