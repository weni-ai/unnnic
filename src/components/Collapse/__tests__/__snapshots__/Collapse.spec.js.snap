// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Collapse.vue > matches the snapshot 1`] = `
"<div data-v-7d0d9fbd="" data-testid="collapse" class="unnnic-collapse unnnic-collapse--size-lg">
  <div data-v-7d0d9fbd="" data-testid="collapse-header" class="unnnic-collapse__header">Test Title<span data-v-26446d8e="" data-v-7d0d9fbd="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--neutral-cloudy unnnic-icon-size--xs unnnic-icon__size--xs" data-testid="material-icon" translate="no">expand_more</span></div>
  <div data-v-7d0d9fbd="" class="unnnic-collapse__body" style="display: none;"></div>
</div>"
`;
